Pipe a context through the CSS serialization code

weinig opened the pull request:
#### 9e3372cb622a8d4497cbe9d1d5c6341869f48af6
<pre>
Pipe a context through the CSS serialization code
<a href="https://bugs.webkit.org/show_bug.cgi?id=287219">https://bugs.webkit.org/show_bug.cgi?id=287219</a>

Reviewed by Simon Fraser.

Adds a new type, CSS::SerializationContext and pipes it through
CSS serialization code.

Initially, this allows removing the set/clear replacement URL
functions as well as some unnecessary StyleProperties copies,
by instead having the replacement URL state on the new context.
This is an important step to removing mutability from CSS values.

Eventually, this will also allow passing down things like the
stage of cascade so we can differentiate serialization of
specified vs. computed styles, and potentially even a builder.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/animation/WebAnimation.cpp:
* Source/WebCore/css/CSSAppleColorFilterPropertyValue.cpp:
* Source/WebCore/css/CSSAppleColorFilterPropertyValue.h:
* Source/WebCore/css/CSSAspectRatioValue.cpp:
* Source/WebCore/css/CSSAspectRatioValue.h:
* Source/WebCore/css/CSSAttrValue.cpp:
* Source/WebCore/css/CSSAttrValue.h:
* Source/WebCore/css/CSSBackgroundRepeatValue.cpp:
* Source/WebCore/css/CSSBackgroundRepeatValue.h:
* Source/WebCore/css/CSSBasicShapeValue.cpp:
* Source/WebCore/css/CSSBasicShapeValue.h:
* Source/WebCore/css/CSSBorderImageSliceValue.cpp:
* Source/WebCore/css/CSSBorderImageSliceValue.h:
* Source/WebCore/css/CSSBorderImageWidthValue.cpp:
* Source/WebCore/css/CSSBorderImageWidthValue.h:
* Source/WebCore/css/CSSBoxShadowPropertyValue.cpp:
* Source/WebCore/css/CSSBoxShadowPropertyValue.h:
* Source/WebCore/css/CSSCanvasValue.cpp:
* Source/WebCore/css/CSSCanvasValue.h:
* Source/WebCore/css/CSSColorSchemeValue.cpp:
* Source/WebCore/css/CSSColorSchemeValue.h:
* Source/WebCore/css/CSSColorValue.cpp:
* Source/WebCore/css/CSSColorValue.h:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
* Source/WebCore/css/CSSContentDistributionValue.cpp:
* Source/WebCore/css/CSSContentDistributionValue.h:
* Source/WebCore/css/CSSCounterValue.cpp:
* Source/WebCore/css/CSSCounterValue.h:
* Source/WebCore/css/CSSCrossfadeValue.cpp:
* Source/WebCore/css/CSSCrossfadeValue.h:
* Source/WebCore/css/CSSCursorImageValue.cpp:
* Source/WebCore/css/CSSCursorImageValue.h:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/CSSEasingFunctionValue.cpp:
* Source/WebCore/css/CSSEasingFunctionValue.h:
* Source/WebCore/css/CSSFilterImageValue.cpp:
* Source/WebCore/css/CSSFilterImageValue.h:
* Source/WebCore/css/CSSFilterPropertyValue.cpp:
* Source/WebCore/css/CSSFilterPropertyValue.h:
* Source/WebCore/css/CSSFontFaceRule.cpp:
* Source/WebCore/css/CSSFontFaceRule.h:
* Source/WebCore/css/CSSFontFaceSrcValue.cpp:
* Source/WebCore/css/CSSFontFaceSrcValue.h:
* Source/WebCore/css/CSSFontFeatureValue.cpp:
* Source/WebCore/css/CSSFontFeatureValue.h:
* Source/WebCore/css/CSSFontPaletteValuesOverrideColorsValue.cpp:
* Source/WebCore/css/CSSFontPaletteValuesOverrideColorsValue.h:
* Source/WebCore/css/CSSFontStyleRangeValue.cpp:
* Source/WebCore/css/CSSFontStyleRangeValue.h:
* Source/WebCore/css/CSSFontStyleWithAngleValue.cpp:
* Source/WebCore/css/CSSFontStyleWithAngleValue.h:
* Source/WebCore/css/CSSFontValue.cpp:
* Source/WebCore/css/CSSFontValue.h:
* Source/WebCore/css/CSSFontVariantAlternatesValue.cpp:
* Source/WebCore/css/CSSFontVariantAlternatesValue.h:
* Source/WebCore/css/CSSFontVariationValue.cpp:
* Source/WebCore/css/CSSFontVariationValue.h:
* Source/WebCore/css/CSSFunctionValue.cpp:
* Source/WebCore/css/CSSFunctionValue.h:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSGradientValue.h:
* Source/WebCore/css/CSSGridAutoRepeatValue.cpp:
* Source/WebCore/css/CSSGridAutoRepeatValue.h:
* Source/WebCore/css/CSSGridIntegerRepeatValue.cpp:
* Source/WebCore/css/CSSGridIntegerRepeatValue.h:
* Source/WebCore/css/CSSGridLineNamesValue.cpp:
* Source/WebCore/css/CSSGridLineNamesValue.h:
* Source/WebCore/css/CSSGridLineValue.cpp:
* Source/WebCore/css/CSSGridLineValue.h:
* Source/WebCore/css/CSSGridTemplateAreasValue.cpp:
* Source/WebCore/css/CSSGridTemplateAreasValue.h:
* Source/WebCore/css/CSSGroupingRule.cpp:
* Source/WebCore/css/CSSGroupingRule.h:
* Source/WebCore/css/CSSImageSetOptionValue.cpp:
* Source/WebCore/css/CSSImageSetOptionValue.h:
* Source/WebCore/css/CSSImageSetValue.cpp:
* Source/WebCore/css/CSSImageSetValue.h:
* Source/WebCore/css/CSSImageValue.cpp:
* Source/WebCore/css/CSSImageValue.h:
* Source/WebCore/css/CSSImportRule.cpp:
* Source/WebCore/css/CSSImportRule.h:
* Source/WebCore/css/CSSKeyframeRule.cpp:
* Source/WebCore/css/CSSLineBoxContainValue.cpp:
* Source/WebCore/css/CSSLineBoxContainValue.h:
* Source/WebCore/css/CSSMediaRule.cpp:
* Source/WebCore/css/CSSMediaRule.h:
* Source/WebCore/css/CSSNamedImageValue.cpp:
* Source/WebCore/css/CSSNamedImageValue.h:
* Source/WebCore/css/CSSNestedDeclarations.cpp:
* Source/WebCore/css/CSSOffsetRotateValue.cpp:
* Source/WebCore/css/CSSOffsetRotateValue.h:
* Source/WebCore/css/CSSPageRule.cpp:
* Source/WebCore/css/CSSPaintImageValue.cpp:
* Source/WebCore/css/CSSPaintImageValue.h:
* Source/WebCore/css/CSSPathValue.cpp:
* Source/WebCore/css/CSSPathValue.h:
* Source/WebCore/css/CSSPendingSubstitutionValue.h:
* Source/WebCore/css/CSSPositionTryRule.cpp:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSQuadValue.cpp:
* Source/WebCore/css/CSSQuadValue.h:
* Source/WebCore/css/CSSRayValue.cpp:
* Source/WebCore/css/CSSRayValue.h:
* Source/WebCore/css/CSSRectValue.cpp:
* Source/WebCore/css/CSSRectValue.h:
* Source/WebCore/css/CSSReflectValue.cpp:
* Source/WebCore/css/CSSReflectValue.h:
* Source/WebCore/css/CSSRule.h:
* Source/WebCore/css/CSSScrollValue.cpp:
* Source/WebCore/css/CSSScrollValue.h:
* Source/WebCore/css/CSSStyleRule.cpp:
* Source/WebCore/css/CSSStyleRule.h:
* Source/WebCore/css/CSSStyleSheet.cpp:
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/css/CSSSubgridValue.cpp:
* Source/WebCore/css/CSSSubgridValue.h:
* Source/WebCore/css/CSSSupportsRule.cpp:
* Source/WebCore/css/CSSSupportsRule.h:
* Source/WebCore/css/CSSTextShadowPropertyValue.cpp:
* Source/WebCore/css/CSSTextShadowPropertyValue.h:
* Source/WebCore/css/CSSTransformListValue.h:
* Source/WebCore/css/CSSUnicodeRangeValue.cpp:
* Source/WebCore/css/CSSUnicodeRangeValue.h:
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/CSSValueList.cpp:
* Source/WebCore/css/CSSValueList.h:
* Source/WebCore/css/CSSValuePair.cpp:
* Source/WebCore/css/CSSValuePair.h:
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
* Source/WebCore/css/CSSVariableReferenceValue.h:
* Source/WebCore/css/CSSViewValue.cpp:
* Source/WebCore/css/CSSViewValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/DeprecatedCSSOMBoxShadowValue.cpp:
* Source/WebCore/css/DeprecatedCSSOMFilterFunctionValue.cpp:
* Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.cpp:
* Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.h:
* Source/WebCore/css/DeprecatedCSSOMTextShadowValue.cpp:
* Source/WebCore/css/DeprecatedCSSOMValue.cpp:
* Source/WebCore/css/DeprecatedCSSOMValue.h:
* Source/WebCore/css/MutableStyleProperties.cpp:
* Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp:
* Source/WebCore/css/Quad.h:
* Source/WebCore/css/Rect.h:
* Source/WebCore/css/ShorthandSerializer.cpp:
* Source/WebCore/css/ShorthandSerializer.h:
* Source/WebCore/css/StyleProperties.cpp:
* Source/WebCore/css/StyleProperties.h:
* Source/WebCore/css/StylePropertiesInlines.h:
* Source/WebCore/css/StyleRule.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.h:
* Source/WebCore/css/calc/CSSCalcTree.cpp:
* Source/WebCore/css/calc/CSSCalcValue.cpp:
* Source/WebCore/css/calc/CSSCalcValue.h:
* Source/WebCore/css/query/GenericMediaQuerySerialization.cpp:
* Source/WebCore/css/typedom/CSSStyleImageValue.cpp:
* Source/WebCore/css/typedom/CSSStyleValue.cpp:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
* Source/WebCore/css/typedom/DeclaredStylePropertyMap.cpp:
* Source/WebCore/css/typedom/InlineStylePropertyMap.cpp:
* Source/WebCore/css/values/CSSSerializationContext.cpp: Added.
* Source/WebCore/css/values/CSSSerializationContext.h: Added.
* Source/WebCore/css/values/CSSValueTypes.cpp:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/css/values/backgrounds/CSSBorderRadius.cpp:
* Source/WebCore/css/values/backgrounds/CSSBorderRadius.h:
* Source/WebCore/css/values/color-adjust/CSSColorScheme.cpp:
* Source/WebCore/css/values/color-adjust/CSSColorScheme.h:
* Source/WebCore/css/values/color/CSSAbsoluteColor.h:
* Source/WebCore/css/values/color/CSSAbsoluteColorSerialization.h:
* Source/WebCore/css/values/color/CSSColor.cpp:
* Source/WebCore/css/values/color/CSSColor.h:
* Source/WebCore/css/values/color/CSSColorLayers.cpp:
* Source/WebCore/css/values/color/CSSColorLayers.h:
* Source/WebCore/css/values/color/CSSColorLayersSerialization.h:
* Source/WebCore/css/values/color/CSSColorMix.cpp:
* Source/WebCore/css/values/color/CSSColorMix.h:
* Source/WebCore/css/values/color/CSSColorMixSerialization.cpp:
* Source/WebCore/css/values/color/CSSColorMixSerialization.h:
* Source/WebCore/css/values/color/CSSContrastColor.cpp:
* Source/WebCore/css/values/color/CSSContrastColor.h:
* Source/WebCore/css/values/color/CSSContrastColorSerialization.h:
* Source/WebCore/css/values/color/CSSHexColor.cpp:
* Source/WebCore/css/values/color/CSSHexColor.h:
* Source/WebCore/css/values/color/CSSKeywordColor.cpp:
* Source/WebCore/css/values/color/CSSKeywordColor.h:
* Source/WebCore/css/values/color/CSSLightDarkColor.cpp:
* Source/WebCore/css/values/color/CSSLightDarkColor.h:
* Source/WebCore/css/values/color/CSSRelativeColor.h:
* Source/WebCore/css/values/color/CSSRelativeColorSerialization.h:
* Source/WebCore/css/values/color/CSSResolvedColor.cpp:
* Source/WebCore/css/values/color/CSSResolvedColor.h:
* Source/WebCore/css/values/easing/CSSStepsEasingFunction.h:
* Source/WebCore/css/values/filter-effects/CSSFilterReference.cpp:
* Source/WebCore/css/values/filter-effects/CSSFilterReference.h:
* Source/WebCore/css/values/images/CSSGradient.cpp:
* Source/WebCore/css/values/images/CSSGradient.h:
* Source/WebCore/css/values/motion/CSSRayFunction.cpp:
* Source/WebCore/css/values/motion/CSSRayFunction.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Serialization.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Serialization.h:
* Source/WebCore/css/values/primitives/CSSSymbol.cpp:
* Source/WebCore/css/values/primitives/CSSSymbol.h:
* Source/WebCore/css/values/primitives/CSSUnevaluatedCalc.cpp:
* Source/WebCore/css/values/primitives/CSSUnevaluatedCalc.h:
* Source/WebCore/css/values/shapes/CSSCircleFunction.cpp:
* Source/WebCore/css/values/shapes/CSSCircleFunction.h:
* Source/WebCore/css/values/shapes/CSSEllipseFunction.cpp:
* Source/WebCore/css/values/shapes/CSSEllipseFunction.h:
* Source/WebCore/css/values/shapes/CSSInsetFunction.cpp:
* Source/WebCore/css/values/shapes/CSSInsetFunction.h:
* Source/WebCore/css/values/shapes/CSSPathFunction.cpp:
* Source/WebCore/css/values/shapes/CSSPathFunction.h:
* Source/WebCore/css/values/shapes/CSSPolygonFunction.cpp:
* Source/WebCore/css/values/shapes/CSSPolygonFunction.h:
* Source/WebCore/css/values/shapes/CSSRectFunction.cpp:
* Source/WebCore/css/values/shapes/CSSRectFunction.h:
* Source/WebCore/css/values/shapes/CSSShapeFunction.cpp:
* Source/WebCore/css/values/shapes/CSSShapeFunction.h:
* Source/WebCore/css/values/shapes/CSSXywhFunction.cpp:
* Source/WebCore/css/values/shapes/CSSXywhFunction.h:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/dom/Element.cpp:
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ExtensionStyleSheets.cpp:
* Source/WebCore/dom/ExtensionStyleSheets.h:
* Source/WebCore/dom/StyledElement.cpp:
* Source/WebCore/dom/StyledElement.h:
* Source/WebCore/editing/ApplyStyleCommand.cpp:
* Source/WebCore/editing/EditingStyle.cpp:
* Source/WebCore/editing/Editor.cpp:
* Source/WebCore/editing/MarkupAccumulator.cpp:
* Source/WebCore/editing/MarkupAccumulator.h:
* Source/WebCore/editing/ReplaceSelectionCommand.cpp:
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
* Source/WebCore/editing/markup.cpp:
* Source/WebCore/editing/markup.h:
* Source/WebCore/html/HTMLImageElement.cpp:
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/html/HTMLSourceElement.cpp:
* Source/WebCore/html/HTMLSourceElement.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
* Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
* Source/WebCore/html/parser/HTMLSrcsetParser.h:
* Source/WebCore/html/track/WebVTTParser.cpp:
* Source/WebCore/inspector/InspectorOverlay.cpp:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
* Source/WebCore/style/values/color/StyleColor.cpp:
* Source/WebCore/style/values/color/StyleColor.h:
* Source/WebCore/style/values/color/StyleColorLayers.cpp:
* Source/WebCore/style/values/color/StyleColorLayers.h:
* Source/WebCore/style/values/color/StyleColorMix.cpp:
* Source/WebCore/style/values/color/StyleColorMix.h:
* Source/WebCore/style/values/color/StyleContrastColor.cpp:
* Source/WebCore/style/values/color/StyleContrastColor.h:
* Source/WebCore/style/values/color/StyleCurrentColor.cpp:
* Source/WebCore/style/values/color/StyleCurrentColor.h:
* Source/WebCore/style/values/color/StyleRelativeColor.h:
* Source/WebCore/style/values/color/StyleResolvedColor.cpp:
* Source/WebCore/style/values/color/StyleResolvedColor.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.h:
* Source/WebCore/svg/properties/SVGPropertyAnimator.h:
* Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp:

Canonical link: <a href="https://commits.webkit.org/290141@main">https://commits.webkit.org/290141@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e6e319587d4fab9b7ad08aac3582233f87a7ff7a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89084 "65 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8608 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/43862 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94062 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/39845 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91135 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/8995 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/16794 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68639 "Failure limit exceed. At least found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/forms/constraints/infinite_backtracking.html (failure)") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26312 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92086 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/6881 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/80813 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49001 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6630 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/35199 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/38952 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/76993 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36172 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/95897 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16264 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/11906 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77517 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16520 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/76612 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76805 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18934 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21206 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/19795 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9356 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16278 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21589 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16019 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19470 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/17800 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/animation/KeyframeEffect.cpp:
```diff
@@ -920,7 +921,7 @@ auto KeyframeEffect::getKeyframes() -> Vector<ComputedKeyframe>
                 if (cssPropertyId == CSSPropertyCustom)
                     continue;
                 if (auto cssValue = parsedKeyframe.style->getPropertyCSSValue(cssPropertyId))
-                    stringValue = cssValue->cssText();
+                    stringValue = cssValue->cssText(CSS::defaultSerializationContext());
```
I think the reason this function has the strangely terse and vague name `cssText` is that we wanted to exactly match the DOM API back when this was web exposed. As this gets further and further away from something web exposed, I’d be tempted to give it a name with the word _serialize_ in it and consider not including the word CSS. After all this is a CSS value, so what would be the non-CSS text? If we are touching every single call site, it seems like we may want to rethink the name.

weinig replied:
This isn't the last time I will touch all these call sites. But trying to make too many changes here at once has been a downfall of mine in my last few attempts at doing this.

In addition to renaming, I think these should all take StringBuilder& instead of returning a string. Alternatively, we can put a StringBuilder on the SerializationContext to keep the number of parameters low.

I think `serialize()` is a good name.

darinadler commented on Source/WebCore/css/CSSImageValue.cpp:
```diff
@@ -123,39 +124,25 @@ bool CSSImageValue::customTraverseSubresources(const Function<bool(const CachedR
 
-    if (!m_replacementURLString.isEmpty())
-        return serializeURL(m_replacementURLString);
+    if (auto replacementURLString = context.replacementURLStrings.get(m_location.resolvedURL.string()); !replacementURLString.isEmpty())
```
Is `m_location.resolvedURL.string()` guaranteed to not be a null string?

weinig replied:
I would assume so as the previous code did:

```c++
auto replacementURLString = replacementURLStrings.get(m_location.resolvedURL.string());
```

without any null check, but I have not proved it true for myself.

darinadler commented on Source/WebCore/css/CSSValue.h:
```diff
@@ -137,8 +141,6 @@ class CSSValue {
     // FIXME: These three traversing functions are buggy. It should be rewritten with visitChildren.
```
Not three any more!

weinig replied:
Oops -> https://github.com/WebKit/WebKit/pull/40359 !

weinig committed:
Pipe a context through the CSS serialization code
https://bugs.webkit.org/show_bug.cgi?id=287219

Reviewed by Simon Fraser.

Adds a new type, CSS::SerializationContext and pipes it through
CSS serialization code.

Initially, this allows removing the set/clear replacement URL
functions as well as some unnecessary StyleProperties copies,
by instead having the replacement URL state on the new context.
This is an important step to removing mutability from CSS values.

Eventually, this will also allow passing down things like the
stage of cascade so we can differentiate serialization of
specified vs. computed styles, and potentially even a builder.

* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/animation/WebAnimation.cpp:
* Source/WebCore/css/CSSAppleColorFilterPropertyValue.cpp:
* Source/WebCore/css/CSSAppleColorFilterPropertyValue.h:
* Source/WebCore/css/CSSAspectRatioValue.cpp:
* Source/WebCore/css/CSSAspectRatioValue.h:
* Source/WebCore/css/CSSAttrValue.cpp:
* Source/WebCore/css/CSSAttrValue.h:
* Source/WebCore/css/CSSBackgroundRepeatValue.cpp:
* Source/WebCore/css/CSSBackgroundRepeatValue.h:
* Source/WebCore/css/CSSBasicShapeValue.cpp:
* Source/WebCore/css/CSSBasicShapeValue.h:
* Source/WebCore/css/CSSBorderImageSliceValue.cpp:
* Source/WebCore/css/CSSBorderImageSliceValue.h:
* Source/WebCore/css/CSSBorderImageWidthValue.cpp:
* Source/WebCore/css/CSSBorderImageWidthValue.h:
* Source/WebCore/css/CSSBoxShadowPropertyValue.cpp:
* Source/WebCore/css/CSSBoxShadowPropertyValue.h:
* Source/WebCore/css/CSSCanvasValue.cpp:
* Source/WebCore/css/CSSCanvasValue.h:
* Source/WebCore/css/CSSColorSchemeValue.cpp:
* Source/WebCore/css/CSSColorSchemeValue.h:
* Source/WebCore/css/CSSColorValue.cpp:
* Source/WebCore/css/CSSColorValue.h:
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
* Source/WebCore/css/CSSContentDistributionValue.cpp:
* Source/WebCore/css/CSSContentDistributionValue.h:
* Source/WebCore/css/CSSCounterValue.cpp:
* Source/WebCore/css/CSSCounterValue.h:
* Source/WebCore/css/CSSCrossfadeValue.cpp:
* Source/WebCore/css/CSSCrossfadeValue.h:
* Source/WebCore/css/CSSCursorImageValue.cpp:
* Source/WebCore/css/CSSCursorImageValue.h:
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/CSSEasingFunctionValue.cpp:
* Source/WebCore/css/CSSEasingFunctionValue.h:
* Source/WebCore/css/CSSFilterImageValue.cpp:
* Source/WebCore/css/CSSFilterImageValue.h:
* Source/WebCore/css/CSSFilterPropertyValue.cpp:
* Source/WebCore/css/CSSFilterPropertyValue.h:
* Source/WebCore/css/CSSFontFaceRule.cpp:
* Source/WebCore/css/CSSFontFaceRule.h:
* Source/WebCore/css/CSSFontFaceSrcValue.cpp:
* Source/WebCore/css/CSSFontFaceSrcValue.h:
* Source/WebCore/css/CSSFontFeatureValue.cpp:
* Source/WebCore/css/CSSFontFeatureValue.h:
* Source/WebCore/css/CSSFontPaletteValuesOverrideColorsValue.cpp:
* Source/WebCore/css/CSSFontPaletteValuesOverrideColorsValue.h:
* Source/WebCore/css/CSSFontStyleRangeValue.cpp:
* Source/WebCore/css/CSSFontStyleRangeValue.h:
* Source/WebCore/css/CSSFontStyleWithAngleValue.cpp:
* Source/WebCore/css/CSSFontStyleWithAngleValue.h:
* Source/WebCore/css/CSSFontValue.cpp:
* Source/WebCore/css/CSSFontValue.h:
* Source/WebCore/css/CSSFontVariantAlternatesValue.cpp:
* Source/WebCore/css/CSSFontVariantAlternatesValue.h:
* Source/WebCore/css/CSSFontVariationValue.cpp:
* Source/WebCore/css/CSSFontVariationValue.h:
* Source/WebCore/css/CSSFunctionValue.cpp:
* Source/WebCore/css/CSSFunctionValue.h:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSGradientValue.h:
* Source/WebCore/css/CSSGridAutoRepeatValue.cpp:
* Source/WebCore/css/CSSGridAutoRepeatValue.h:
* Source/WebCore/css/CSSGridIntegerRepeatValue.cpp:
* Source/WebCore/css/CSSGridIntegerRepeatValue.h:
* Source/WebCore/css/CSSGridLineNamesValue.cpp:
* Source/WebCore/css/CSSGridLineNamesValue.h:
* Source/WebCore/css/CSSGridLineValue.cpp:
* Source/WebCore/css/CSSGridLineValue.h:
* Source/WebCore/css/CSSGridTemplateAreasValue.cpp:
* Source/WebCore/css/CSSGridTemplateAreasValue.h:
* Source/WebCore/css/CSSGroupingRule.cpp:
* Source/WebCore/css/CSSGroupingRule.h:
* Source/WebCore/css/CSSImageSetOptionValue.cpp:
* Source/WebCore/css/CSSImageSetOptionValue.h:
* Source/WebCore/css/CSSImageSetValue.cpp:
* Source/WebCore/css/CSSImageSetValue.h:
* Source/WebCore/css/CSSImageValue.cpp:
* Source/WebCore/css/CSSImageValue.h:
* Source/WebCore/css/CSSImportRule.cpp:
* Source/WebCore/css/CSSImportRule.h:
* Source/WebCore/css/CSSKeyframeRule.cpp:
* Source/WebCore/css/CSSLineBoxContainValue.cpp:
* Source/WebCore/css/CSSLineBoxContainValue.h:
* Source/WebCore/css/CSSMediaRule.cpp:
* Source/WebCore/css/CSSMediaRule.h:
* Source/WebCore/css/CSSNamedImageValue.cpp:
* Source/WebCore/css/CSSNamedImageValue.h:
* Source/WebCore/css/CSSNestedDeclarations.cpp:
* Source/WebCore/css/CSSOffsetRotateValue.cpp:
* Source/WebCore/css/CSSOffsetRotateValue.h:
* Source/WebCore/css/CSSPageRule.cpp:
* Source/WebCore/css/CSSPaintImageValue.cpp:
* Source/WebCore/css/CSSPaintImageValue.h:
* Source/WebCore/css/CSSPathValue.cpp:
* Source/WebCore/css/CSSPathValue.h:
* Source/WebCore/css/CSSPendingSubstitutionValue.h:
* Source/WebCore/css/CSSPositionTryRule.cpp:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSQuadValue.cpp:
* Source/WebCore/css/CSSQuadValue.h:
* Source/WebCore/css/CSSRayValue.cpp:
* Source/WebCore/css/CSSRayValue.h:
* Source/WebCore/css/CSSRectValue.cpp:
* Source/WebCore/css/CSSRectValue.h:
* Source/WebCore/css/CSSReflectValue.cpp:
* Source/WebCore/css/CSSReflectValue.h:
* Source/WebCore/css/CSSRule.h:
* Source/WebCore/css/CSSScrollValue.cpp:
* Source/WebCore/css/CSSScrollValue.h:
* Source/WebCore/css/CSSStyleRule.cpp:
* Source/WebCore/css/CSSStyleRule.h:
* Source/WebCore/css/CSSStyleSheet.cpp:
* Source/WebCore/css/CSSStyleSheet.h:
* Source/WebCore/css/CSSSubgridValue.cpp:
* Source/WebCore/css/CSSSubgridValue.h:
* Source/WebCore/css/CSSSupportsRule.cpp:
* Source/WebCore/css/CSSSupportsRule.h:
* Source/WebCore/css/CSSTextShadowPropertyValue.cpp:
* Source/WebCore/css/CSSTextShadowPropertyValue.h:
* Source/WebCore/css/CSSTransformListValue.h:
* Source/WebCore/css/CSSUnicodeRangeValue.cpp:
* Source/WebCore/css/CSSUnicodeRangeValue.h:
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/CSSValueList.cpp:
* Source/WebCore/css/CSSValueList.h:
* Source/WebCore/css/CSSValuePair.cpp:
* Source/WebCore/css/CSSValuePair.h:
* Source/WebCore/css/CSSVariableReferenceValue.cpp:
* Source/WebCore/css/CSSVariableReferenceValue.h:
* Source/WebCore/css/CSSViewValue.cpp:
* Source/WebCore/css/CSSViewValue.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/DeprecatedCSSOMBoxShadowValue.cpp:
* Source/WebCore/css/DeprecatedCSSOMFilterFunctionValue.cpp:
* Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.cpp:
* Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.h:
* Source/WebCore/css/DeprecatedCSSOMTextShadowValue.cpp:
* Source/WebCore/css/DeprecatedCSSOMValue.cpp:
* Source/WebCore/css/DeprecatedCSSOMValue.h:
* Source/WebCore/css/MutableStyleProperties.cpp:
* Source/WebCore/css/PropertySetCSSStyleDeclaration.cpp:
* Source/WebCore/css/Quad.h:
* Source/WebCore/css/Rect.h:
* Source/WebCore/css/ShorthandSerializer.cpp:
* Source/WebCore/css/ShorthandSerializer.h:
* Source/WebCore/css/StyleProperties.cpp:
* Source/WebCore/css/StyleProperties.h:
* Source/WebCore/css/StylePropertiesInlines.h:
* Source/WebCore/css/StyleRule.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Parser.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.h:
* Source/WebCore/css/calc/CSSCalcTree.cpp:
* Source/WebCore/css/calc/CSSCalcValue.cpp:
* Source/WebCore/css/calc/CSSCalcValue.h:
* Source/WebCore/css/query/GenericMediaQuerySerialization.cpp:
* Source/WebCore/css/typedom/CSSStyleImageValue.cpp:
* Source/WebCore/css/typedom/CSSStyleValue.cpp:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
* Source/WebCore/css/typedom/ComputedStylePropertyMapReadOnly.cpp:
* Source/WebCore/css/typedom/DeclaredStylePropertyMap.cpp:
* Source/WebCore/css/typedom/InlineStylePropertyMap.cpp:
* Source/WebCore/css/values/CSSSerializationContext.cpp: Added.
* Source/WebCore/css/values/CSSSerializationContext.h: Added.
* Source/WebCore/css/values/CSSValueTypes.cpp:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/css/values/backgrounds/CSSBorderRadius.cpp:
* Source/WebCore/css/values/backgrounds/CSSBorderRadius.h:
* Source/WebCore/css/values/color-adjust/CSSColorScheme.cpp:
* Source/WebCore/css/values/color-adjust/CSSColorScheme.h:
* Source/WebCore/css/values/color/CSSAbsoluteColor.h:
* Source/WebCore/css/values/color/CSSAbsoluteColorSerialization.h:
* Source/WebCore/css/values/color/CSSColor.cpp:
* Source/WebCore/css/values/color/CSSColor.h:
* Source/WebCore/css/values/color/CSSColorLayers.cpp:
* Source/WebCore/css/values/color/CSSColorLayers.h:
* Source/WebCore/css/values/color/CSSColorLayersSerialization.h:
* Source/WebCore/css/values/color/CSSColorMix.cpp:
* Source/WebCore/css/values/color/CSSColorMix.h:
* Source/WebCore/css/values/color/CSSColorMixSerialization.cpp:
* Source/WebCore/css/values/color/CSSColorMixSerialization.h:
* Source/WebCore/css/values/color/CSSContrastColor.cpp:
* Source/WebCore/css/values/color/CSSContrastColor.h:
* Source/WebCore/css/values/color/CSSContrastColorSerialization.h:
* Source/WebCore/css/values/color/CSSHexColor.cpp:
* Source/WebCore/css/values/color/CSSHexColor.h:
* Source/WebCore/css/values/color/CSSKeywordColor.cpp:
* Source/WebCore/css/values/color/CSSKeywordColor.h:
* Source/WebCore/css/values/color/CSSLightDarkColor.cpp:
* Source/WebCore/css/values/color/CSSLightDarkColor.h:
* Source/WebCore/css/values/color/CSSRelativeColor.h:
* Source/WebCore/css/values/color/CSSRelativeColorSerialization.h:
* Source/WebCore/css/values/color/CSSResolvedColor.cpp:
* Source/WebCore/css/values/color/CSSResolvedColor.h:
* Source/WebCore/css/values/easing/CSSStepsEasingFunction.h:
* Source/WebCore/css/values/filter-effects/CSSFilterReference.cpp:
* Source/WebCore/css/values/filter-effects/CSSFilterReference.h:
* Source/WebCore/css/values/images/CSSGradient.cpp:
* Source/WebCore/css/values/images/CSSGradient.h:
* Source/WebCore/css/values/motion/CSSRayFunction.cpp:
* Source/WebCore/css/values/motion/CSSRayFunction.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Serialization.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+Serialization.h:
* Source/WebCore/css/values/primitives/CSSSymbol.cpp:
* Source/WebCore/css/values/primitives/CSSSymbol.h:
* Source/WebCore/css/values/primitives/CSSUnevaluatedCalc.cpp:
* Source/WebCore/css/values/primitives/CSSUnevaluatedCalc.h:
* Source/WebCore/css/values/shapes/CSSCircleFunction.cpp:
* Source/WebCore/css/values/shapes/CSSCircleFunction.h:
* Source/WebCore/css/values/shapes/CSSEllipseFunction.cpp:
* Source/WebCore/css/values/shapes/CSSEllipseFunction.h:
* Source/WebCore/css/values/shapes/CSSInsetFunction.cpp:
* Source/WebCore/css/values/shapes/CSSInsetFunction.h:
* Source/WebCore/css/values/shapes/CSSPathFunction.cpp:
* Source/WebCore/css/values/shapes/CSSPathFunction.h:
* Source/WebCore/css/values/shapes/CSSPolygonFunction.cpp:
* Source/WebCore/css/values/shapes/CSSPolygonFunction.h:
* Source/WebCore/css/values/shapes/CSSRectFunction.cpp:
* Source/WebCore/css/values/shapes/CSSRectFunction.h:
* Source/WebCore/css/values/shapes/CSSShapeFunction.cpp:
* Source/WebCore/css/values/shapes/CSSShapeFunction.h:
* Source/WebCore/css/values/shapes/CSSXywhFunction.cpp:
* Source/WebCore/css/values/shapes/CSSXywhFunction.h:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/dom/Element.cpp:
* Source/WebCore/dom/Element.h:
* Source/WebCore/dom/ExtensionStyleSheets.cpp:
* Source/WebCore/dom/ExtensionStyleSheets.h:
* Source/WebCore/dom/StyledElement.cpp:
* Source/WebCore/dom/StyledElement.h:
* Source/WebCore/editing/ApplyStyleCommand.cpp:
* Source/WebCore/editing/EditingStyle.cpp:
* Source/WebCore/editing/Editor.cpp:
* Source/WebCore/editing/MarkupAccumulator.cpp:
* Source/WebCore/editing/MarkupAccumulator.h:
* Source/WebCore/editing/ReplaceSelectionCommand.cpp:
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
* Source/WebCore/editing/markup.cpp:
* Source/WebCore/editing/markup.h:
* Source/WebCore/html/HTMLImageElement.cpp:
* Source/WebCore/html/HTMLImageElement.h:
* Source/WebCore/html/HTMLSourceElement.cpp:
* Source/WebCore/html/HTMLSourceElement.h:
* Source/WebCore/html/canvas/CanvasRenderingContext2DBase.cpp:
* Source/WebCore/html/parser/HTMLSrcsetParser.cpp:
* Source/WebCore/html/parser/HTMLSrcsetParser.h:
* Source/WebCore/html/track/WebVTTParser.cpp:
* Source/WebCore/inspector/InspectorOverlay.cpp:
* Source/WebCore/inspector/agents/InspectorAnimationAgent.cpp:
* Source/WebCore/loader/archive/cf/LegacyWebArchive.cpp:
* Source/WebCore/style/values/color/StyleColor.cpp:
* Source/WebCore/style/values/color/StyleColor.h:
* Source/WebCore/style/values/color/StyleColorLayers.cpp:
* Source/WebCore/style/values/color/StyleColorLayers.h:
* Source/WebCore/style/values/color/StyleColorMix.cpp:
* Source/WebCore/style/values/color/StyleColorMix.h:
* Source/WebCore/style/values/color/StyleContrastColor.cpp:
* Source/WebCore/style/values/color/StyleContrastColor.h:
* Source/WebCore/style/values/color/StyleCurrentColor.cpp:
* Source/WebCore/style/values/color/StyleCurrentColor.h:
* Source/WebCore/style/values/color/StyleRelativeColor.h:
* Source/WebCore/style/values/color/StyleResolvedColor.cpp:
* Source/WebCore/style/values/color/StyleResolvedColor.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Logging.h:
* Source/WebCore/svg/properties/SVGPropertyAnimator.h:
* Tools/TestWebKitAPI/Tests/WebCore/CSSParser.cpp:

Canonical link: https://commits.webkit.org/290141@main
